src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,254 @@
+   * @return pointer to the first row.
+   */
+  byte *GetPayload() { return varlen_contents_; }
+
+ private:
+  uint32_t checksum_;
+  catalog::table_oid_t table_oid_;
+  byte varlen_contents_[0];
+};
+
+/**
+ * A buffered tuple writer collects tuples into blocks, and output a block once a time.
+ * The block size is fetched from database setting, and should be the page size.
+ * layout:
+ * ----------------------------------------------------------------------------------
+ * | checksum | table_oid | tuple1 | varlen_entries(if exist) | tuple2 | tuple3 | ... |
+ * ----------------------------------------------------------------------------------
+ */
+class BufferedTupleWriter {
+  // TODO(Zhaozhe): checksum
+ public:
+  BufferedTupleWriter() = default;
+
+  /**
+   * Instantiate a new BufferedTupleWriter, open a new checkpoint file to write into.
+   * @param log_file_path path to the checkpoint file.
+   */
+  explicit BufferedTupleWriter(const char *log_file_path)
+      : out_(PosixIoWrappers::Open(log_file_path, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR)),
+        block_size_(CHECKPOINT_BLOCK_SIZE),
+        buffer_(new byte[block_size_]()) {
+    ResetBuffer();
+  }
+
+  /**
+   * Open a file to write into
+   * @param log_file_path path to the checkpoint file.
+   */
+  void Open(const char *log_file_path) {
+    out_ = PosixIoWrappers::Open(log_file_path, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
+    block_size_ = CHECKPOINT_BLOCK_SIZE;
+    buffer_ = new byte[block_size_]();
+    ResetBuffer();
+  }
+
+  /**
+   * Write the content of the buffer into disk.
+   */
+  void Persist() { PersistBuffer(); }
+
+  /**
+   * Close current file and release the buffer.
+   */
+  void Close() {
+    PosixIoWrappers::Close(out_);
+    delete[] buffer_;
+  }
+
+  /**
+   * Serialize the tuple into the buffer, and write to disk when the buffer is full.
+   * @param row pointer to the RowView given by ProjectedColumn
+   * @param row_buffer pointer to the ProjectedRow buffer.
+   * @param layout of the Row.
+   */
+  void SerializeTuple(ProjectedColumns::RowView *row, ProjectedRow *row_buffer, const storage::BlockLayout &layout);
+
+  /**
+   * Serialize a ProjectedRow and its varlens into the file.
+   * @param row_buffer pointer to the row to be serialized.
+   * @param total_varlen total length of all varlens of this row.
+   * @param varlen_entries all the varlen entries in this row.
+   */
+  void AppendTupleToBuffer(ProjectedRow *row_buffer, int32_t total_varlen,
+                           const std::vector<const VarlenEntry *> &varlen_entries);
+
+  /**
+   * Get the current checkpoint page.
+   * @return pointer to the checkpoint page header.
+   */
+  CheckpointFilePage *GetPage() { return reinterpret_cast<CheckpointFilePage *>(buffer_); }
+
+ private:
+  int out_;  // fd of the output files
+  uint32_t block_size_;
+  uint32_t cur_buffer_size_ = 0;
+  byte *buffer_ = nullptr;
+
+  void ResetBuffer() {
+    memset(buffer_, 0, block_size_);
```

Comment:
Why do you need to zero the buffer?  Nothing in your data structures are null-terminated so "uninitialized" data shouldn't be a problem.

Reply:
Thanks. I think just zeroing the checksum bits will be fine.

Reply:
Yes. In the latest version. the only place we need to set to zeros is after the last row in the page to suggest no more rows in this page. We do not need to reset the whole buffer anymore.